Feature/recheck intro (#42)
retest/docs: docs/recheck/introduction/index.md

# Introduction

***recheck*** is an API which implements *Difference Testing* to check objects across multiple versions and collect differences.

It therefore defines a domain model specific to ***recheck*** so that it can do Difference Testing. That means, however, that there needs to be a conversion from the passed object to the domain model. While ***recheck***  itself cannot convert those objects, there is the possibility to use available extensions or expand recheck via own extensions as described under [installation](installation.md).

## Difference Testing

> The idea of *Difference Testing* is derived from the idea of *Golden Master Testing* or *Characterization Testing*. Based on these practices, it describes the complete state of the *System under Test (SUT)*, where the complete state is determined by the particular technology used. By observing said state, it essentially differs from the classical approach to software testing by removing the need for assertions and eliminating the use of specifications. That's because *observing* does not mean *verification*; Difference Testing accepts every output it encounters, saves it and compares upon re-execution.

Reviewer: ```suggestion
The idea of *Difference Testing* is derived from the idea of *Golden Master Testing* (a.k.a. [Characterization Testing](https://en.wikipedia.org/wiki/Characterization_test)). Based on these practices, it describes the complete state of the object under test, where the complete state is determined by extension of the particular technology used. By observing said state, it essentially differs from the classical approach to software testing by removing the need for assertions and eliminating the use of specifications. That's because *observing* does not mean *verification*; Difference Testing accepts every output it encounters, saves it and compares upon re-execution.
```

Difference Testing is basically described with four steps:

> 1. Define a Golden Master based upon an evaluated version of the SUT.

Reviewer: Above you speak about "objects" (I already suggest a change from "system under test" to "object under test"), here you specifically say SUT. I would suggest to be more generic.

2. Convert and save the SUT as a Golden Master by initially executing the test.
3. Compare to a newer version of the SUT by re-executing the test.

> 4. If a failure happens: Verify the encountered differences (by a human).

Reviewer: ```suggestion
4. If differences occur: let a human decide if these diffs should be accepted (update the Golden Master), ignored (update an ignore file similar to `.gitignore`), or rejected (hooray, you've found a bug!).
```

### Advantages

* There is no need to define individual assertions as the state is captured completely.

> * Changed elements can still be found by identifying them with other attributes.

Reviewer: This is very specific to GUIs I think?

Author: Yes that might be, but it is a great selling argument. Furthermore recheck itself supports it, while not as good as the unbreakable, but it tries to find changed elements via the threshold.
